wlutil/br/br.py

```diff
@@ -0,0 +1,107 @@
+import os
+import subprocess as sp
+import shutil
+import logging
+import wlutil
+import string
+
+# Note: All argument paths are expected to be absolute paths
+
+# Some common directories for this module (all absolute paths)
+br_dir = os.path.dirname(os.path.realpath(__file__))
+overlay = os.path.join(br_dir, 'firesim-overlay')
+
+initTemplate = string.Template("""#!/bin/sh
+
+SYSLOGD_ARGS=-n
+KLOGD_ARGS=-n
+
+start() {
+    echo "FIRESIM RUN START" && /firesim.sh $args && echo "FIRESIM RUN END"
+}
+
+case "$$1" in
+  start)
+  start
+  ;;
+  stop)
+  #stop
+  ;;
+  restart|reload)
+  start
+  ;;
+  *)
+  echo "Usage: $$0 {start|stop|restart}"
+  exit 1
+esac
+
+exit""")
+
+class Builder:
+
+    def baseConfig(self):
+        return {
+                'name' : 'buildroot-base',
+                'distro' : 'br',
+                'workdir' : br_dir,
+                'builder' : self,
+                'img' : os.path.join(br_dir, "buildroot/output/images/rootfs.ext2")
+                }
+
+    # Build a base image in the requested format and return an absolute path to that image
+    def buildBaseImage(self):
+        log = logging.getLogger()
+        rootfs_target = "rootfs.img"
+        shutil.copy(os.path.join(br_dir, 'buildroot-config'),
+                    os.path.join(br_dir, "buildroot/.config"))
+        # log.debug(sp.check_output(['make'], cwd=os.path.join(br_dir, "buildroot")))
+
+        # Buildroot complains about some common PERL configurations
+        env = os.environ.copy()
+        env.pop('PERL_MM_OPT', None)
+        wlutil.run(['make'], cwd=os.path.join(br_dir, "buildroot"), env=env)
+
+    # Return True if the base image is up to date, or False if it needs to be
+    # rebuilt.
+    def upToDate(self):
+        # XXX There's something wrong with buildroots makefile, it throws an
+        # error and never reports being up to date.
+        # XXX DONT COMMIT THIS CHANGE YOUR DEFNITELY GOING TO FORGET TO UNDO THIS
```

Comment:
This needs to be dealt with.

Reply:
That comment was from when I had it hard-coded to never even check. Now it checks if there is a buildroot output, if not, it will rebuild. If there is some output, it will use it. This means it won't rebuild buildroot if you change some core thing in the buildroot distribution (like buildroot-config) or the buildroot overlay. To change those things, just run make in wlutil/br. Without this, every buildroot-based workload gets fully reproduced every time (10x+ build time).

The comment has been fixed, and I've added a readme to wlutil/br to clarify the situation. In practice, this isn't really an issue because even init scripts and stuff are handled through firemarshal overlays instead of buildroot overlays. We could (and maybe should at some point) replace buildroot with a pre-built image like we do for Fedora.